### ISSUE TITLE
Configurable HttpOnly Cookie

### ISSUE THREAD

**Issue by itanex** [aspnet/Antiforgery#125]
### Configurable HttpOnly Cookie
Allow API user to Configure if the cookie generated is HttpOnly or not

### Functional impact
* Allows for use of ValidationToken Facilities in Angular and other frameworks
* Changes Hard coded true value to an optional value

### Minimal repro steps
1. Create an Angular Application and try to call endpoints secured with [ValidateToken] or [AutoValidateToken]
2. Angular cannot find HTTPOnly cookies because of their security principals
3. Angular does not properly send the ValidationToken HTTPHeader 
** Example Repo: [Antiforgery Demo](https://github.com/itanex/AntiforgeryDemo)

### Expected result
Expected that because of the Architecture design, a developer using ValidateToken Configured for Angular Cookie and Header functionality, that the ValidationToken Cookie should be configured correctly.

### Actual result
Angular is prevented from reading the ValidationToken Cookie because it is set as HTTPOnly, which prevents angular calling endpoints secured with [ValidateToken] or [AutoValidateToken]

### Further technical details

[Documentation for AngularJS V1](https://docs.angularjs.org/api/ng/service/$http#cross-site-request-forgery-xsrf-protection) CSRF CookieToHeader Architecture 

[Wikipedia CSRF Cookie to Header](https://en.wikipedia.org/wiki/Cross-site_request_forgery#Cookie-to-Header_Token) which states that by the architectural design of the CookieToHeader Architecture the ValidationToken cookie needs to be readable by JavaScript

Root of issue, ellipses for brevity

    namespace Microsoft.AspNetCore.Antiforgery.Internal
    {
        public class DefaultAntiforgeryTokenStore : IAntiforgeryTokenStore
        {
            ...
            
            public void SaveCookieToken(HttpContext httpContext, string token)
            {
                ...
                options.HttpOnly = true; // This is hardcoded and cannot be changed
                ...
            }

            ...
        }
    }

**Comment by chrarnoldus** [comment]
Making the cookie created by the DefaultAntiforgeryTokenStore JavaScript-readable wouldn't help, because the token stored in there (the CookieToken) is different from the one the request is validated against (the RequestToken).

Instead, you should send the RequestToken client side, as shown in the following example: https://github.com/aspnet/Entropy/blob/dev/samples/Antiforgery.Angular1/Startup.cs#L36

**Comment by itanex** [comment]
The example does not support usage to variable routes, The example only creates a cookie when the path is  / or /index.html. This is not reusable in an angular app using Html5mode for its routing.

Additionally, Using my example in conjunction with [EditThisCookie](https://chrome.google.com/webstore/detail/editthiscookie/fngmhnnpilhplaeedifhccceomclgfbg?utm_source=chrome-app-launcher-info-dialog) you can actually change the cookie to be readable, reload the page and than submit a request that is accepted.

**Comment by chrarnoldus** [comment]
In my experience the tokens are different (some information about the user is prepended to the RequestToken), but maybe that's only true when using authentication middleware.

Still, creating a second, JavaScript-readable cookie should solve your problem I think (or sending it to the client by some other means).

**Comment by itanex** [comment]
Having the cookie as HttpOnly does not make it a persistent cookie. HttpOnly makes it so that the cookie only is read by the request handler

A Persistent cookie is a cookie with an expiry date that is not a Session cookie. In the purest sense of the word persistent, "continuing to exist or endure over a prolonged period". 

A Session Cookie is a cookie that is destroyed once a browser session is closed.

> According to the Microsoft Developer Network, HttpOnly is an additional flag included in a Set-Cookie HTTP response header. Using the HttpOnly flag when generating a cookie helps mitigate the risk of client side script accessing the protected cookie (if the browser supports it). - [OWASP](https://www.owasp.org/index.php/HttpOnly)

The example you keep referencing is not designed in such a way that one can develop a service using AngularJS/React and the default behaviors of ValidationToken. What I am labeling as a bug is the fact that as a developer I cannot use ValidationToken with the configurations set up using the AntiForgery Service APIs without massive adjustments to the code of an application.

Providing a Configuration option to set this value to false if desired should be allowed, as to comply with the necessary Architectural needs of these and other JavaScript Development Platforms. Code such as the following would be a welcome addition, as it allows a developer to program the token process accordingly for their needs.

Service Configuration

    services.AddAntiforgery(options => {
        options.CookieName= "XSRF-TOKEN";
        options.HeaderName = "X-XSRF-TOKEN";
        options.IsHttpOnlyCookie= false; // default is true
    });

DefaultAntiforgeryTokenStore Update

    namespace Microsoft.AspNetCore.Antiforgery.Internal
    {
        public class DefaultAntiforgeryTokenStore : IAntiforgeryTokenStore
        {
            ...
            public void SaveCookieToken(HttpContext httpContext, string token)
            {
                ...
                options.HttpOnly = _options.IsHttpOnlyCookie; // Code change suggestion
                ...
            }
            ...
        }
    }

AntiforgeryOptions Update

    namespace Microsoft.AspNetCore.Antiforgery
    {
        public class AntiforgeryOptions
        {
            ...
            public boolean IsHttpOnlyCookie { get; set; }
            ...
        }
    }

**Comment by chrarnoldus** [comment]
> I cannot use ValidationToken with the configurations set up using the AntiForgery Service APIs without massive adjustments to the code of an application.

I don't see how you need massive code adjustments. In your example code, remove (https://github.com/itanex/AntiforgeryDemo/blob/master/src/AntiForgeryDemo/Startup.cs#L34)

```
options.CookieName = "XSRF-TOKEN";
```

and replace (https://github.com/itanex/AntiforgeryDemo/blob/master/src/AntiForgeryDemo/Middleware/AntiForgeryMiddleware.cs#L23)

```
antiforgery.GetAndStoreTokens(context);
```

by

```
var requestToken = antiforgery.GetAndStoreTokens(context).RequestToken;
context.Response.Cookies.Append("XSRF-TOKEN", requestToken, new CookieOptions { HttpOnly = false });
```

and stuff will work. You will have two cookies with different content, so I doubt that making the .AspNetCore.Antiforgery cookie JavaScript-readable (like you propose) would actually help.

**Comment by itanex** [comment]
@chrarnoldus 

    options.CookieName = "XSRF-TOKEN" 

This does nothing but change the AntiForgery Cookie name that is generated. It does not allow AngularJS/React or any other framework to read the cookie.

    antiforgery.GetAndStoreTokens(context);

This creates the tokens and creates the session and cookie. The example code does not allow antiforgery tokens to be handled naturally as the expectation of the library provides. It also creates two cookies. 

The AntiForgery Service should still make the cookie HttpOnly by default, I am not suggesting otherwise. The change I am suggesting makes the AntiForgery Service more usable and hacking solutions like the example presents are no longer needed, by allowing the developer to configure this setting.

**Comment by itanex** [comment]
In tests by me and my team we have discovered that our request is superfluous. As it turns out there were several tests including my own which were at fault for the data that spurred this issue.

After cloning and running several painstakingly detailed step-through passes, the cookie that is generated by AntiForgery is not of the value or purpose we need. We discovered that this cookie is in fact part of the actual Antiforgery validation process. Which our previous tests were persisting instead of destroying during our testing scenarios.

I have updated the [Antiforgery Demo](https://github.com/itanex/AntiforgeryDemo) to include the additional cookie similar to the process that is present in the demo example. In addition
